Fix authorsAlpha

subhramit opened the pull request:
# Fix authorsAlpha
**[Auxiliary PR,  required for the GSoC '24 CSL4LibreOffice project]**

Fix for the function `authorsAlpha` in `BracketedPattern` which returns author representation for the purpose of alphanumeric styles.
Adapted for the purpose of DIN 1505-02.
Changes:
1. Max authors before "exceed limit" is 4.
2. No appending "+" if exceeded
3. For a single author, (maximum) first 4 letters of the author name are taken instead of 3.

Investigation ref. #11577  and issue https://github.com/subhramit/jabref/issues/17


References:
https://forums.zotero.org/discussion/12239/style-request-lecture-notes-in-informatics-lni-from-the-germany-computer-science-society
https://forums.zotero.org/discussion/82660/format-din-1505-2-alphanumeric-german-standard-superseded-by-iso-690
https://forums.zotero.org/discussion/82631/problem-with-citation-style-din-1505-2-alphanumeric-german-and-word-integration
Template attached to first link: [ISSS-word-template.docx](https://github.com/user-attachments/files/16591064/ISSS-word-template.docx)
### Mandatory checks


<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

koppor commented:
The `authorsAlpha` implementation is aligned with "GI e.V."'s biblatexs style: https://github.com/gi-ev/biblatex-lni

Output: 

![image](https://github.com/user-attachments/assets/821db4ca-75ec-4a8f-bfa2-4b0f2ae3b673)

LaTeX-Source: https://github.com/gi-ev/biblatex-lni/blob/main/basic-test-en.tex

If we change something in the behavior, all things need to be aligned.

The "ultimate" truth is the Word template at https://gi.de/service/publikationen/lni. But I remember, that also does not cover all cases.

koppor commented:
The current test cases are:

```java
    static Stream<Arguments> authorsAlpha() {
        return Stream.of(
                Arguments.of("A+", "Alexander Artemenko and others"),
                Arguments.of("A+", "Aachen and others"),
                Arguments.of("AB+", "Aachen and Berlin and others"),
                Arguments.of("ABC+", "Aachen and Berlin and Chemnitz and others"),
                // FIXME: Arguments.of("Aach", "Aachen"),
                Arguments.of("AB", "Aachen and Berlin"),
                Arguments.of("ABC", "Aachen and Berlin and Chemnitz"),
                Arguments.of("ABCD", "Aachen and Berlin and Chemnitz and Düsseldorf"),
                Arguments.of("ABC+", "Aachen and Berlin and Chemnitz and Düsseldorf and others"),
                Arguments.of("ABC+", "Aachen and Berlin and Chemnitz and Düsseldorf and Essen"),
                Arguments.of("ABC+", "Aachen and Berlin and Chemnitz and Düsseldorf and Essen and others"));
    }
```

If the test case do not fail at this PR, you found a simpler implementation.

koppor commented:
The update does not support "and others" any more:

![image](https://github.com/user-attachments/assets/0ca42e25-e489-4f3c-8422-b44c485b5bf0)

subhramit commented:
> The current test cases are:
>                 Arguments.of("ABC+", "Aachen and Berlin and Chemnitz and Düsseldorf and others"),
>                 Arguments.of("ABC+", "Aachen and Berlin and Chemnitz and Düsseldorf and Essen"),
>                 Arguments.of("ABC+", "Aachen and Berlin and Chemnitz and Düsseldorf and Essen and others"));
>     }
> ```
> 
> If the test case do not fail at this PR, you found a simpler implementation.

ABC+ won't work with this one. I had removed the + for DIn 1505-02.
Also, ABC+ implies three max authors represented, while the PR changes it to 4 (ABCD and no +)

koppor commented:
After double check with the LNI guidelines, it is OK to remove the `+`.

For proper software engineering:

- [ ] Add support for handling "and others"  (no "+" but two letters of the first last name)
- [ ] Adapt existing test cases
- [ ] Add LNI test cases to test cases (https://github.com/JabRef/jabref/pull/11614#issuecomment-2284786529)

subhramit commented:
Sorry, I mis-clicked on edit instead of reply.
@koppor 
Could you clarify a bit more on:
> * [ ]  Add support for handling "and others" (no "+" but two letters of the first last name)

Meaning - if there is "and others" in the title, what should be done differently? Should it be treated as if we ignore "and others"? First last name of whom? First author? All authors? And when? (is the two letters thing related to/valid only in the case of "and others"?)
https://forums.zotero.org/discussion/82631/problem-with-citation-style-din-1505-2-alphanumeric-german-and-word-integration has the following:
![image](https://github.com/user-attachments/assets/fd92c065-90e2-42ca-99d3-5f0e5ef464de)

That is, till three authors, a total of 4 letters would be filled, depending on how many "extra authors" are there after 1.

koppor commented:
I thought, the rules for "exceeding" are clear.

Four letters of the first last name.

Treat `and others` as exceeding.

koppor commented:
> [forums.zotero.org/discussion/82631/problem-with-citation-style-din-1505-2-alphanumeric-german-and-word-integration](https://forums.zotero.org/discussion/82631/problem-with-citation-style-din-1505-2-alphanumeric-german-and-word-integration) has

A random dude in the net writing something in a comment without trustable sources. - Moreover, his examples lead to different results as shown at https://github.com/JabRef/jabref/pull/11614#issuecomment-2284786529.

I think, the pattern the linuxguy said is supported by JabRef too, but I don't find at https://docs.jabref.org/setup/citationkeypatterns

subhramit commented:
Side note: Also required for PR-D of CSL4LibreOffice.

subhramit commented:
> After double check with the LNI guidelines, it is OK to remove the +.
> 
> For proper software engineering:
> 
> * [ ]  Add support for handling "and others" (no "+" but two letters of the first last name)


Can you tell me if the following expected outputs are correct?:
```java
Arguments.of("A+", "Alexander Artemenko and others"),
                Arguments.of("Aa", "Aachen and others"),
                Arguments.of("Aa", "Aachen and Berlin and others"),
                Arguments.of("Aa", "Aachen and Berlin and Chemnitz and others"),
                Arguments.of("AB", "Aachen and Berlin"),
                Arguments.of("ABC", "Aachen and Berlin and Chemnitz"),
                Arguments.of("ABCD", "Aachen and Berlin and Chemnitz and Düsseldorf"),
                Arguments.of("Aa", "Aachen and Berlin and Chemnitz and Düsseldorf and others"),
                Arguments.of("ABCD", "Aachen and Berlin and Chemnitz and Düsseldorf and Essen"),
                Arguments.of("Aa", "Aachen and Berlin and Chemnitz and Düsseldorf and Essen and others"));
```

subhramit commented:
> I thought, the rules for "exceeding" are clear.
> 
> Four letters of the first last name.
> 
> Treat `and others` as exceeding.

In the comment stating requirements, you mentioned two letters of the first last name instead of four (as mentioned here). I would request more clarification with examples.

koppor commented:
> > I thought, the rules for "exceeding" are clear.
> > 
> > Four letters of the first last name.
> > 
> > Treat `and others` as exceeding.
> 
> In the comment stating requirements, you mentioned two letters of the first last name instead of four (as mentioned here). I would request more clarification with examples for "and others" case with multiple authors (refer the cases above).

I take https://github.com/JabRef/jabref/pull/11614#issuecomment-2284786529 as ground truth. They use two letters. We follow that.

With "exceeding" I meant, it exceeds maximum number of authors. Thus, "and others" means five authors present. Thus, it is case 1.

subhramit commented:
Also, the last two examples from the template have:
```
[Wa14a] Wasser, K.; Feuer, H.; Erde, R.; Licht, H.: Essenzen der Informatik. Verlag Formvoll, 2014.
[Wa14b] Wasser, K.; Feuer, H.; Erde, R.; Licht, H.: Ganz neue Essenzen der Informatik im selben Jahr. Format-Verlag, 2014.
```
[Wa14a] and [Wa14b] imply there has to be an alphabetical counter that gets auto-incremented if two instances of authors are exactly the same but the papers are different. CSL uses itemDataProvider, which automatically supplies numbers, which are then taken care of by reference marks. We would need a similar system for this - but also that detects difference only in papers with the same authors. Furthermore, the letter has to be appended after the "Year" part, which is not handled by authorsAlpha.

Is this a compulsory requirement for this PR? Or should we just add the other cases to the test case and keep this as an edge case and try to think of it later?

subhramit commented:
> Also, the last two examples from the template have:
> 
> ```
> [Wa14a] Wasser, K.; Feuer, H.; Erde, R.; Licht, H.: Essenzen der Informatik. Verlag Formvoll, 2014.
> [Wa14b] Wasser, K.; Feuer, H.; Erde, R.; Licht, H.: Ganz neue Essenzen der Informatik im selben Jahr. Format-Verlag, 2014.
> ```
> 
> [Wa14a] and [Wa14b] imply there has to be an alphabetical counter that gets auto-incremented if two instances of authors are exactly the same but the papers are different. CSL uses itemDataProvider, which automatically supplies numbers, which are then taken care of by reference marks. We would need a similar system for this - but also that detects difference only in papers with the same authors. Furthermore, the letter has to be appended after the "Year" part, which is not handled by authorsAlpha.
> 
> Is this a compulsory requirement for this PR? Or should we just add the other cases to the test case and keep this as an edge case and try to think of it later?

This is also slightly contradictory, as it does not exceed the maximum number of authors (=4), yet uses the first author's two characters only instead of the first character from all 4.

subhramit commented:
> Also, the last two examples from the template have:
> 
> ```
> [Wa14a] Wasser, K.; Feuer, H.; Erde, R.; Licht, H.: Essenzen der Informatik. Verlag Formvoll, 2014.
> [Wa14b] Wasser, K.; Feuer, H.; Erde, R.; Licht, H.: Ganz neue Essenzen der Informatik im selben Jahr. Format-Verlag, 2014.
> ```
> 
> [Wa14a] and [Wa14b] imply there has to be an alphabetical counter that gets auto-incremented if two instances of authors are exactly the same but the papers are different. CSL uses itemDataProvider, which automatically supplies numbers, which are then taken care of by reference marks. We would need a similar system for this - but also that detects difference only in papers with the same authors. Furthermore, the letter has to be appended after the "Year" part, which is not handled by authorsAlpha.
> 
> Is this a compulsory requirement for this PR? Or should we just add the other cases to the test case and keep this as an edge case and try to think of it later?

Okay, after some thought, I am skipping these last two cases. Reason: It would also require "future insight" - if the same authors are going to be cited again in future, we have to append an "a" at the end of the first instance right away. Since this will not be possible, the only way is to have targeted editing of the citations as more such entries are cited, which is a hard problem again (auto-refreshment of old citations).

koppor commented:
> Is this a compulsory requirement for this PR? Or should we just add the other cases to the test case and keep this as an edge case and try to think of it later?

One by one. Keep focus. Divide and conquer.

"Just" fix authorsAlpha.

---

JabRef has this functionality internally already. Check the preferences of the citation key generator.

Maybe, it can be reused. Maybe only if a complete bibdatabasecontext is available. Then, the code needs to be refactored.

subhramit commented:
@koppor please see if the changes are fine.
Only case I did not handle are:
```
[An14] 	Anteil an Frauen in der Informatik. Statistics Worldwide, 2014.
```
(since it did not have any name in it, I figured by translating, nor commas that would separate a name like `GI, Gesellschaft für Informatik e.V.`, which is handled). This seems to be out of scope for "authorsAlpha".

Is this OK for now?

subhramit commented:
Merge conflicts...why?
Nevertheless, this is the final version from my side.
Many more test cases covered, much less code.

koppor commented:
> Merge conflicts...why?

I spend 10 minutes to "debug"

When merging commit 8bf6f0523702a68268dcbced120425c55214757d into here, there is a conflict. This commit is from your PR #11577.

The diff is

```
-                                                     .collect(Collectors.toList());
+                                                     .toList();
```

(Side note: The IntelliJ "Resolve Conflicts" dialog is not helpful:

![image](https://github.com/user-attachments/assets/005eebe6-9830-4143-97f9-fc3f2420a3f5)

---

In future, please start coding on an updated `main` branch.

I checked https://devdocs.jabref.org/contributing#contribute-code. We do not say anything about how to branch off `upstream/main`. The command line is 

- `git fetch upstream`
- `git checkout -b <new-branch-name> upstream/main`

Maybe this should be added to the contributors guide?

Side note: Therefore, `git town` introduced [`git hack <branch>`](https://www.git-town.com/commands/hack). I think, the above two commands are better than requiring users to install another `git` too.

---

> Nevertheless, this is the final version from my side. Many more test cases covered, much less code.

I will resolve the merge conflicts for you and then check. Note that tests are NOT run on the CI if there are merge conflicts.

subhramit commented:
> I checked https://devdocs.jabref.org/contributing#contribute-code. We do not say anything about how to branch off `upstream/main`. The command line is
> 
> * `git fetch upstream`
> * `git checkout -b <new-branch-name> upstream/main`
> 
> Maybe this should be added to the contributors guide?

The other day, @Siedlerchr too suggested that it is easier for maintainers to review and merge PRs if we branch off `upstream/main`. I was thinking of cloning directly and adding a branch from main, but I was wondering if contributors can safely do that as it was missing from devdocs. (https://devdocs.jabref.org/contributing.html#understanding-the-basics-of-code-contributions).

If it's just these two commands, I will file a PR adding it to the devdocs. Should we "prefer" new contributors to branch from main as well? Or should it be mentioned as Choice 1/ Choice 2 in the docs?

koppor commented:
> The other day, @Siedlerchr too suggested that it is easier for maintainers to review and merge PRs if we branch off `upstream/main`. I was thinking of cloning directly and adding a branch from main, but I was wondering if contributors can safely do that as it was missing from devdocs ([devdocs.jabref.org/contributing.html#understanding-the-basics-of-code-contributions](https://devdocs.jabref.org/contributing.html#understanding-the-basics-of-code-contributions)).

I highlight the current text regarding the branches.

![image](https://github.com/user-attachments/assets/2268ae2e-9f29-4dc7-9c32-1ba5dd76ff99)

I think, points 2 and 3 need to be merged. And refined with the two git commands.

> If it's just these two commands, I will file a PR adding it to the devdocs. Should we "prefer" new contributors to branch from main as well? Or should it just be mentioned as Choice 1/ Choice 2 in the docs?

There is no choice. It is always `upstream/main`. I have no clue, what some other choice could be...

subhramit commented:
> There is no choice. It is always `upstream/main`. I have no clue, what some other choice could be...

My bad, that was unclear. I meant branching from the fork (origin) vs the main repo (upstream). I also saw Loay mention that builds are generated for branches from upstream.

subhramit commented:
Summary of last two commits (resolution of failing unit tests):
1. Removed "+" from the end of expected strings in `CitationKeyGeneratorTest` for the tests involving `authorsAlpha`.
2. Adapted expected authorsAlpha output "vdAal" to "Aa" for input "Wil van der Aalst" and so on, as per the convention we are following this PR onwards.
3. Changed number of expected characters from authorsAlpha output e.g. from "New" to "Ne" for input "Isaac Newton".
4. Similar changes in `BracketedPatternTest`.
5. [Additional] Removed Exceptions that were not being thrown from tests in `AbstractCitationKeyPatternsTest` and `BibEntryTest` along the way.

koppor commented on src/main/java/org/jabref/logic/citationkeypattern/BracketedPattern.java:
```diff
@@ -846,51 +846,36 @@ static String allAuthors(AuthorList authorList) {
         return alphaStyle.toString();
     }
 
+    static String getLastName(Author author) {
+        String[] nameParts = author.getNamePrefixAndFamilyName()
+                                   .replaceAll("\\s+", " ")
+                                   .trim()
+                                   .split(" ");
+        return nameParts[nameParts.length - 1];
+    }
```
Why can't you use `org.jabref.model.entry.Author#getFamilyName`?

subhramit replied:
Just tried it. The tests fail.

subhramit replied:
Changed.

subhramit commented on src/test/java/org/jabref/logic/citationkeypattern/BracketedPatternTest.java:
```diff
@@ -76,22 +77,65 @@ void allAuthors(String expected, AuthorList list) {
-                Arguments.of("ABC+", "Aachen and Berlin and Chemnitz and Düsseldorf and Essen"),
-                Arguments.of("ABC+", "Aachen and Berlin and Chemnitz and Düsseldorf and Essen and others"));
+                Arguments.of("Aa", "Aachen and Berlin and Chemnitz and Düsseldorf and others"),
+                Arguments.of("ABCD", "Aachen and Berlin and Chemnitz and Düsseldorf and Essen"),
```
As per our discussion, we would do that if there is "and others".

subhramit replied:
This has been established ever since we started talking about DIN. I can do only the first author's two letters of last name in case of exceeding authors but note that everywhere I have looked, in case of exceeding authors (and not "and others"), the abbreviation of first four authors are taken.
You can try citing 10.3945/ajcn.111.023457 using citationmachine as well: https://www.citationmachine.net/bibliographies/a446c586-d900-4583-a249-72d7b9bc0b52
![image](https://github.com/user-attachments/assets/b8b803ae-b7fa-40a4-8749-633c9d451983)

koppor replied:
Besides the LNI template: https://github.com/JabRef/jabref/pull/11614#issuecomment-2284786529

We need to setup a minimal latex+bibtex document with authorsalpha with the test cases and see what that returns.

koppor commented:
@subhramit It is so great to see that now two persons (you and me) are in the context of `authorsalpha`.

Can you add a test for institutions?

In BibTeX, they are written as `author = {{Gesellschaft für Informatik e.V.`}}`. In JabRef then `{Gesellschaft für Informatik e.V.}` (because the outer braces are removed while reading (and added while writing)).

I am curious whether the thing works there, too.

koppor commented:
It is good enough for now.

What I understood:

- DIN is outdated
- DIN is requested in the Zotero forum because of LNI use
- LNI publishes its guidelines
- LNI differs from DIN in some cases

What I miss:

- other appearances in the wild of that style (besides "my" scientific-thesis-tempalte: https://github.com/latextemplates/scientific-thesis-template
- minimal example using bibtex authorsalpha
- a structured markdown file where the sources of the decisions are listed (and also a TL;DR summary is given) -- "keygenerators.md" in folder https://github.com/JabRef/jabref/tree/main/docs/code-howtos  is a good place

Reason for the last point: We exchange information at multiple different places (which makes it scattered), but the core knowledge is not condensed. What if someone looks into that in one year - and if none of us both is available? --> Documentaiton enables asynchronous work.

Follow-up TODOs:

- [ ] Collect knowledge in `keygenerators.md`.  (New PR!)
- [ ] Create `.tex` and `.bib` file for `authorsalpha` and the different cases (New PR)
- [ ] Find usages of `authorsalpha` in the wild. (Add it to `keygenerators.md`.  (New PR)
- [ ] check institutions - example: `author = {{Apache Foundation}}`. It is OK if it fails, we just need to know and then think of solutions. For BibLaTeX, there is the feature `label = {AF}`. (add result to `keygenerators.md` - New PR). Add test case for that. Maybe commented out or marked with `// TODO ...`

koppor commented on src/main/java/org/jabref/logic/citationkeypattern/BracketedPattern.java:
```diff
@@ -868,14 +868,6 @@ public static String authorsAlpha(AuthorList authorList) {
-                                   .trim()
-                                   .split(" ");
-        return nameParts[nameParts.length - 1];
-    }
```
Code hint - if you encounter a similar case in the future, there is a more intuitive code IMHO:

https://www.w3schools.com/java/ref_string_lastindexof.asp

```java
int pos = string.lastIndexOf(" ");
String result = string.substr(pos +1);
```

subhramit commented:
> It is good enough for now.
> 
> What I understood:
> 
> * DIN is outdated
> * DIN is requested in the Zotero forum because of LNI use
> * LNI publishes its guidelines
> * LNI differs from DIN in some cases
> 
> What I miss:
> 
> * other appearances in the wild of that style (besides "my" scientific-thesis-tempalte: https://github.com/latextemplates/scientific-thesis-template
> * minimal example using bibtex authorsalpha
> * a structured markdown file where the sources of the decisions are listed (and also a TL;DR summary is given) -- "keygenerators.md" in folder https://github.com/JabRef/jabref/tree/main/docs/code-howtos  is a good place
> 
> Reason for the last point: We exchange information at multiple different places (which makes it scattered), but the core knowledge is not condensed. What if someone looks into that in one year - and if none of us both is available? --> Documentaiton enables asynchronous work.
> 
> Follow-up TODOs:
> 
> * [ ]  Collect knowledge in `keygenerators.md`.  (New PR!)
> * [ ]  Create `.tex` and `.bib` file for `authorsalpha` and the different cases (New PR)
> * [ ]  Find usages of `authorsalpha` in the wild. (Add it to `keygenerators.md`.  (New PR)
> * [x]  check institutions - example: `author = {{Apache Foundation}}`. It is OK if it fails, we just need to know and then think of solutions. For BibLaTeX, there is the feature `label = {AF}`. (add result to `keygenerators.md` - New PR). Add test case for that. Maybe commented out or marked with `// TODO ...`

I have handled the case of institutions now.

koppor commented on src/main/java/org/jabref/logic/citationkeypattern/BracketedPattern.java:
```diff
@@ -846,51 +846,45 @@ static String allAuthors(AuthorList authorList) {
+        if (numberOfAuthors == 1 || andOthersPresent) {
+            // Single author or "and others" case
+            String lastName = authorList.getAuthor(0).getFamilyName().orElse("");
+            if (lastName.startsWith("{") && lastName.endsWith("}")) {
+                alphaStyle.append(getOrganizationInitials(lastName));
```
Rewrite using org.jabref.logic.formatter.bibtexfields.RemoveEnclosingBracesFormatter

1. format the string using that formatter
2. if result is not equal -> you found an organization
3. Pass that string to `getOrganizationInitials` - the first two lines can be removed then.

subhramit replied:
Done.

koppor reviewed (approved):
Seems to be better than before 😅.

Follow-ups as discussed.

koppor commented:
> * [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at [JabRef/user-documentation/issues](https://github.com/JabRef/user-documentation/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen) or, even better, I submitted a pull request to the documentation repository.

Where is the issue? There was no link in the PR description here - and at the issue tracker, I only found https://github.com/JabRef/user-documentation/issues/453

subhramit commented:
> Where is the issue? There was no link in the PR description here - and at the issue tracker, I only found [JabRef/user-documentation#453](https://github.com/JabRef/user-documentation/issues/453)

The issue is mentioned in the PR description:
> Investigation ref. https://github.com/JabRef/jabref/pull/11577 and issue https://github.com/subhramit/jabref/issues/17

I am/was not aware of any other existing issues around this. We started brainstorming when the requirement for adding alphanumeric style support in CSL came up (in the above pull/issue).
